controller/controlplane/controller_reconciler_utils.go

```diff
@@ -120,3 +123,127 @@ func (r *Reconciler) ensureAdminMTLSCertificateSecret(
 		matchingLabels,
 	)
 }
+
+func (r *Reconciler) validateWatchNamespaceGrants(
+	ctx context.Context,
+	cp *ControlPlane,
+) ([]string, error) {
+	if cp.Spec.WatchNamespaces == nil {
+		return nil, errors.New("spec.watchNamespaces cannot be empty")
+	}
+
+	switch cp.Spec.WatchNamespaces.Type {
+	// NOTE: We currentlty do not require any ReferenceGrants or other permission
+	// granting resources for the "All" case.
+	case operatorv2alpha1.WatchNamespacesTypeAll:
+		return nil, nil
+	// No special permissions are required to watch the controlplane's own namespace.
+	case operatorv2alpha1.WatchNamespacesTypeOwn:
+		return []string{cp.Namespace}, nil
+	case operatorv2alpha1.WatchNamespacesTypeList:
+		var nsList []string
+		for _, ns := range cp.Spec.WatchNamespaces.List {
+			if err := ensureWatchNamespaceGrantsForNamespace(ctx, r.Client, cp, ns); err != nil {
+				return nsList, err
+			}
+			nsList = append(nsList, ns)
+		}
+		// Add ControlPlane's own namespace as it will add it anyway because
+		// that's where the default "publish service" exists.
+		// We add it here as we do not require a ReferenceGrant for own namespace
+		// so there's no validation whether a grant exists.
+		nsList = append(nsList, cp.Namespace)
+
+		return nsList, nil
+	default:
+		return nil, fmt.Errorf("unexpected watchNamespaces.type: %q", cp.Spec.WatchNamespaces.Type)
+	}
+}
+
+// +kubebuilder:rbac:groups=gateway-operator.konghq.com,resources=watchnamespacegrants,verbs=list
+
+// ensureWatchNamespaceGrantsForNamespace ensures that a WatchNamespaceGrant exists for the
+// given namespace and ControlPlane.
+// It returns an error if a WatchNamespaceGrant is missing.
+func ensureWatchNamespaceGrantsForNamespace(
+	ctx context.Context,
+	cl client.Client,
+	cp *ControlPlane,
+	ns string,
+) error {
+	var grants operatorv1alpha1.WatchNamespaceGrantList
+	if err := cl.List(ctx, &grants, client.InNamespace(ns)); err != nil {
+		return fmt.Errorf("failed listing WatchNamespaceGrants in namespace %s: %w", ns, err)
+	}
+	for _, grant := range grants.Items {
+		if !watchNamespaceGrantContainsControlPlaneFrom(grant, cp) {
```

Comment:
Does this mean "If one item contains the grant to watch the namespace, return nil?" If so I think structure like this should be clearer

```suggestion
		if watchNamespaceGrantContainsControlPlaneFrom(grant, cp) {
		// return nil if there is one grant allows the CP to watch the namespace.
		    return nil
		}
```